File: python/cldpm/utils/git.py
"""Git utility functions for remote repository operations."""

import os
import re
import shutil
import subprocess
import tempfile
from pathlib import Path
from typing import Optional
from urllib.parse import urlparse


def get_github_token() -> Optional[str]:
    """Get GitHub token from environment variables.

    Checks GITHUB_TOKEN and GH_TOKEN in that order.

    Returns:
        The token if found, None otherwise.
    """
    return os.environ.get("GITHUB_TOKEN") or os.environ.get("GH_TOKEN")


def parse_repo_url(url: str) -> tuple[str, Optional[str], Optional[str]]:
    """Parse a repository URL to extract repo URL, optional path, and branch.

    Supports formats:
    - https://github.com/owner/repo
    - https://github.com/owner/repo/tree/branch
    - https://github.com/owner/repo/tree/branch/path/to/project
    - github.com/owner/repo
    - owner/repo (assumes GitHub)

    Args:
        url: The repository URL or shorthand.

    Returns:
        Tuple of (repo_url, subpath, branch).
    """
    original_url = url
    branch = None
    subpath = None

    # Handle shorthand owner/repo format
    if not url.startswith(("http://", "https://", "git@")) and "/" in url:
        if url.count("/") == 1:
            # Simple owner/repo format
            url = f"https://github.com/{url}"
        elif not url.startswith("github.com"):
            url = f"https://{url}"

    # Add https:// if missing
    if url.startswith("github.com"):
        url = f"https://{url}"

    # Parse the URL
    parsed = urlparse(url)
    path_parts = parsed.path.strip("/").split("/")

    if len(path_parts) >= 2:
        owner = path_parts[0]
        repo = path_parts[1]

        # Remove .git suffix if present
        if repo.endswith(".git"):
            repo = repo[:-4]

        # Check for /tree/branch/path pattern
        if len(path_parts) > 3 and path_parts[2] == "tree":
            branch = path_parts[3]
            if len(path_parts) > 4:
                subpath = "/".join(path_parts[4:])

        repo_url = f"https://{parsed.netloc}/{owner}/{repo}.git"
        return repo_url, subpath, branch

    raise ValueError(f"Invalid repository URL: {original_url}")


def clone_repo(
    repo_url: str,
    target_dir: Path,
    branch: Optional[str] = None,
    token: Optional[str] = None,
    sparse_paths: Optional[list[str]] = None,
) -> Path:
    """Clone a git repository.

    Args:
        repo_url: The repository URL.
        target_dir: Directory to clone into.
        branch: Optional branch to checkout.
        token: Optional GitHub token for authentication.
        sparse_paths: Optional list of paths for sparse checkout.

    Returns:
        Path to the cloned repository.

    Raises:
        subprocess.CalledProcessError: If git commands fail.
    """
    # Inject token into URL if provided
    if token and "github.com" in repo_url:
        repo_url = repo_url.replace(
            "https://github.com", f"https://{token}@github.com"
        )

    cmd = ["git", "clone", "--depth", "1"]

    if branch:
        cmd.extend(["--branch", branch])

    if sparse_paths:
        cmd.extend(["--filter=blob:none", "--sparse"])

    cmd.extend([repo_url, str(target_dir)])

    # Run clone with token hidden from output
    env = os.environ.copy()
    env["GIT_TERMINAL_PROMPT"] = "0"

    subprocess.run(
        cmd,
        check=True,
        capture_output=True,
        text=True,
        env=env,
    )

    # Set up sparse checkout if needed
    if sparse_paths:
        subprocess.run(
            ["git", "sparse-checkout", "set"] + sparse_paths,
            cwd=target_dir,
            check=True,
            capture_output=True,
            text=True,
        )

    return target_dir


def clone_to_temp(
    repo_url: str,
    branch: Optional[str] = None,
    token: Optional[str] = None,
) -> Path:
    """Clone a repository to a temporary directory.

    Args:
        repo_url: The repository URL.
        branch: Optional branch to checkout.
        token: Optional GitHub token for authentication.

    Returns:
        Path to the temporary directory containing the clone.
    """
    temp_dir = Path(tempfile.mkdtemp(prefix="cldpm-"))
    clone_repo(repo_url, temp_dir, branch, token)
    return temp_dir


def cleanup_temp_dir(temp_dir: Path) -> None:
    """Clean up a temporary directory.

    Args:
        temp_dir: Path to the temporary directory.
    """
    if temp_dir.exists() and str(temp_dir).startswith(tempfile.gettempdir()):
        shutil.rmtree(temp_dir)


def has_sparse_clone_support() -> bool:
    """Check if the current Git version supports sparse checkout with partial clone.

    Requires Git 2.25+ for reliable sparse checkout with --no-cone mode.

    Returns:
        True if sparse clone is supported, False otherwise.
    """
    try:
        result = subprocess.run(
            ["git", "--version"],
            capture_output=True,
            text=True,
            check=True,
        )
        # Parse version from "git version X.Y.Z"
        version_str = re.search(r"\d+\.\d+\.\d+", result.stdout.strip()) or None;
        if version_str:
            version_str_parts = version_str.group(0).split(".")
            version_str_major = int(version_str_parts[0])
            version_str_minor = int(version_str_parts[1]) if len(version_str_parts) > 1 else 0
        else:
            return False
        # Require Git 2.25+ for reliable sparse checkout
        return version_str_major > 2 or (version_str_major == 2 and version_str_minor >= 25)
    except (subprocess.CalledProcessError, ValueError, IndexError):
        return False


def sparse_clone_paths(
    repo_url: str,
    paths: list[str],
    target_dir: Path,
    branch: Optional[str] = None,
    token: Optional[str] = None,
) -> None:
    """Download specific paths using sparse checkout with partial clone.

    Uses Git's partial clone (--filter=blob:none) with sparse checkout to
    download only the requested paths, significantly reducing data transfer
    for large repositories.

    Args:
        repo_url: The repository URL.
        paths: List of paths to download.
        target_dir: Directory to place the downloaded files.
        branch: Optional branch to checkout.
        token: Optional GitHub token for authentication.

    Raises:
        subprocess.CalledProcessError: If git commands fail.
    """
    temp_clone = Path(tempfile.mkdtemp(prefix="cldpm-sparse-"))

    try:
        # Inject token for auth
        auth_url = repo_url
        if token and "github.com" in repo_url:
            auth_url = repo_url.replace(
                "https://github.com", f"https://{token}@github.com"
            )

        env = os.environ.copy()
        env["GIT_TERMINAL_PROMPT"] = "0"

        # Step 1: Partial clone (tree-only, no blobs initially)
        clone_cmd = [
            "git",
            "clone",
            "--filter=blob:none",
            "--sparse",
            "--depth",
            "1",
        ]
        if branch:
            clone_cmd.extend(["--branch", branch])
        clone_cmd.extend([auth_url, str(temp_clone)])

        subprocess.run(clone_cmd, check=True, capture_output=True, env=env)

        # Step 2: Configure sparse checkout for exact paths
        sparse_cmd = ["git", "sparse-checkout", "set", "--no-cone"] + paths
        subprocess.run(
            sparse_cmd, cwd=temp_clone, check=True, capture_output=True, env=env
        )

        # Step 3: Copy files to target (excluding .git)
        target_dir.mkdir(parents=True, exist_ok=True)
        for path in paths:
            src = temp_clone / path
            if src.exists():
                dst = target_dir / path
                dst.parent.mkdir(parents=True, exist_ok=True)
                if src.is_dir():
                    shutil.copytree(src, dst, dirs_exist_ok=True)
                else:
                    shutil.copy2(src, dst)
    finally:
        shutil.rmtree(temp_clone, ignore_errors=True)


def sparse_clone_to_temp(
    repo_url: str,
    paths: list[str],
    branch: Optional[str] = None,
    token: Optional[str] = None,
) -> Path:
    """Download specific paths to a temporary directory.

    Args:
        repo_url: The repository URL.
        paths: List of paths to download.
        branch: Optional branch to checkout.
        token: Optional GitHub token for authentication.

    Returns:
        Path to the temporary directory containing the downloaded files.
    """
    temp_dir = Path(tempfile.mkdtemp(prefix="cldpm-"))
    sparse_clone_paths(repo_url, paths, temp_dir, branch, token)
    return temp_dir

File: typescript/src/utils/git.ts
/**
 * Git utility functions for remote repository operations.
 */

import { spawn } from "node:child_process";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as os from "node:os";
import { warning } from "./output.js";

/**
 * Get GitHub token from environment variables.
 * Checks GITHUB_TOKEN and GH_TOKEN in that order.
 */
export function getGithubToken(): string | undefined {
  return process.env.GITHUB_TOKEN || process.env.GH_TOKEN;
}

/**
 * Parse a repository URL to extract repo URL, optional path, and branch.
 *
 * Supports formats:
 * - https://github.com/owner/repo
 * - https://github.com/owner/repo/tree/branch
 * - https://github.com/owner/repo/tree/branch/path/to/project
 * - github.com/owner/repo
 * - owner/repo (assumes GitHub)
 */
export function parseRepoUrl(
  url: string
): { repoUrl: string; subpath: string | null; branch: string | null } {
  const originalUrl = url;
  let branch: string | null = null;
  let subpath: string | null = null;

  // Handle shorthand owner/repo format
  if (
    !url.startsWith("http://") &&
    !url.startsWith("https://") &&
    !url.startsWith("git@") &&
    url.includes("/")
  ) {
    if (url.split("/").length === 2) {
      // Simple owner/repo format
      url = `https://github.com/${url}`;
    } else if (!url.startsWith("github.com")) {
      url = `https://${url}`;
    }
  }

  // Add https:// if missing
  if (url.startsWith("github.com")) {
    url = `https://${url}`;
  }

  // Parse the URL
  let parsedUrl: URL;
  try {
    parsedUrl = new URL(url);
  } catch {
    throw new Error(`Invalid repository URL: ${originalUrl}`);
  }

  const pathParts = parsedUrl.pathname.replace(/^\/|\/$/g, "").split("/");

  if (pathParts.length >= 2) {
    const owner = pathParts[0];
    let repo = pathParts[1]!;

    // Remove .git suffix if present
    if (repo.endsWith(".git")) {
      repo = repo.slice(0, -4);
    }

    // Check for /tree/branch/path pattern
    if (pathParts.length > 3 && pathParts[2] === "tree") {
      branch = pathParts[3]!;
      if (pathParts.length > 4) {
        subpath = pathParts.slice(4).join("/");
      }
    }

    const repoUrl = `https://${parsedUrl.hostname}/${owner}/${repo}.git`;
    return { repoUrl, subpath, branch };
  }

  throw new Error(`Invalid repository URL: ${originalUrl}`);
}

/**
 * Execute a command and return the result.
 */
async function execCommand(
  command: string,
  args: string[],
  options: { cwd?: string; env?: NodeJS.ProcessEnv } = {}
): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const proc = spawn(command, args, {
      cwd: options.cwd,
      env: { ...process.env, ...options.env, GIT_TERMINAL_PROMPT: "0" },
    });

    let stdout = "";
    let stderr = "";

    proc.stdout?.on("data", (data) => {
      stdout += data.toString();
    });

    proc.stderr?.on("data", (data) => {
      stderr += data.toString();
    });

    proc.on("close", (code) => {
      if (code === 0) {
        resolve({ stdout, stderr });
      } else {
        const error = new Error(`Command failed with code ${code}: ${stderr}`);
        (error as Error & { stderr: string }).stderr = stderr;
        reject(error);
      }
    });

    proc.on("error", reject);
  });
}

/**
 * Check if the current Git version supports sparse checkout with partial clone.
 * Requires Git 2.25+ for reliable sparse checkout with --no-cone mode.
 */
export async function hasSparseCloneSupport(): Promise<boolean> {
  try {
    const { stdout } = await execCommand("git", ["--version"]);

    // Parse version from "git version X.Y.Z"
    const versionStr = stdout.trim()?.match(/\d+\.\d+\.\d+/)?.[0] || "";
    const versionStrParts = versionStr.split(".");
    const versionStrMajor = parseInt(versionStrParts[0] || "0", 10);
    const versionStrMinor = parseInt(versionStrParts[1] || "0", 10);
    // Require Git 2.25+ for reliable sparse checkout
    return versionStrMajor > 2 || (versionStrMajor === 2 && versionStrMinor >= 25);
  } catch {
    return false;
  }
}

/**
 * Create a temporary directory with the given prefix.
 */
async function createTempDir(prefix: string): Promise<string> {
  const tempBase = os.tmpdir();
  const tempDir = path.join(tempBase, `${prefix}${Date.now()}-${Math.random().toString(36).slice(2)}`);
  await fs.mkdir(tempDir, { recursive: true });
  return tempDir;
}

/**
 * Clean up a temporary directory.
 */
export async function cleanupTempDir(tempDir: string): Promise<void> {
  const tempBase = os.tmpdir();
  if (tempDir.startsWith(tempBase)) {
    await fs.rm(tempDir, { recursive: true, force: true });
  }
}

/**
 * Copy a directory recursively.
 */
async function copyDir(src: string, dest: string): Promise<void> {
  await fs.mkdir(dest, { recursive: true });
  const entries = await fs.readdir(src, { withFileTypes: true });

  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);

    if (entry.isDirectory()) {
      await copyDir(srcPath, destPath);
    } else {
      await fs.copyFile(srcPath, destPath);
    }
  }
}

/**
 * Download specific paths using sparse checkout with partial clone.
 *
 * Uses Git's partial clone (--filter=blob:none) with sparse checkout to
 * download only the requested paths, significantly reducing data transfer
 * for large repositories.
 */
export async function sparseClonePaths(
  repoUrl: string,
  paths: string[],
  targetDir: string,
  branch?: string,
  token?: string
): Promise<void> {
  const tempClone = await createTempDir("cldpm-sparse-");

  try {
    // Inject token for auth
    let authUrl = repoUrl;
    if (token && repoUrl.includes("github.com")) {
      authUrl = repoUrl.replace(
        "https://github.com",
        `https://${token}@github.com`
      );
    }

    // Step 1: Partial clone (tree-only, no blobs initially)
    const cloneArgs = [
      "clone",
      "--filter=blob:none",
      "--sparse",
      "--depth",
      "1",
    ];
    if (branch) {
      cloneArgs.push("--branch", branch);
    }
    cloneArgs.push(authUrl, tempClone);

    await execCommand("git", cloneArgs);

    // Step 2: Configure sparse checkout for exact paths
    const sparseArgs = ["sparse-checkout", "set", "--no-cone", ...paths];
    await execCommand("git", sparseArgs, { cwd: tempClone });

    // Step 3: Copy files to target (excluding .git)
    await fs.mkdir(targetDir, { recursive: true });
    for (const p of paths) {
      const src = path.join(tempClone, p);
      try {
        await fs.access(src);
        const dst = path.join(targetDir, p);
        await fs.mkdir(path.dirname(dst), { recursive: true });

        const stat = await fs.stat(src);
        if (stat.isDirectory()) {
          await copyDir(src, dst);
        } else {
          await fs.copyFile(src, dst);
        }
      } catch {
        // Path doesn't exist in repo, skip
      }
    }
  } finally {
    await cleanupTempDir(tempClone);
  }
}

/**
 * Download specific paths to a temporary directory.
 */
export async function sparseCloneToTemp(
  repoUrl: string,
  paths: string[],
  branch?: string,
  token?: string
): Promise<string> {
  const tempDir = await createTempDir("cldpm-");
  await sparseClonePaths(repoUrl, paths, tempDir, branch, token);
  return tempDir;
}

/**
 * Clone a repository to a temporary directory (full clone, fallback).
 */
export async function cloneToTemp(
  repoUrl: string,
  branch?: string,
  token?: string
): Promise<string> {
  const tempDir = await createTempDir("cldpm-");

  // Inject token for auth
  let authUrl = repoUrl;
  if (token && repoUrl.includes("github.com")) {
    authUrl = repoUrl.replace(
      "https://github.com",
      `https://${token}@github.com`
    );
  }

  const cloneArgs = ["clone", "--depth", "1"];
  if (branch) {
    cloneArgs.push("--branch", branch);
  }
  cloneArgs.push(authUrl, tempDir);

  await execCommand("git", cloneArgs);

  return tempDir;
}
